### PR TITLE
4.x email renderer

### PR DESCRIPTION
**Pull request by ADmad**
Refs #12114.

### CODE REVIEW

File: src/Mailer/Email.php [hunk 1]
```diff
@@ -250,6 +233,13 @@ class Email implements JsonSerializable, Serializable
      */
     protected $transferEncoding;
 
+    /**
+     * Email Renderer
+     *
+     * @var \Cake\Mailer\Renderer
```
Comment (dereuromark) [inline src/Mailer/Email.php]: |null

**Review by dereuromark — APPROVED** [review]
Apart from my comment above

File: src/Mailer/Renderer.php [hunk 2]
```diff
@@ -0,0 +1,462 @@
...
+ */
+namespace Cake\Mailer;
+
+use Cake\Core\Configure;
+use Cake\Http\Client\FormDataPart;
+use Cake\Utility\Hash;
+use Cake\Utility\Security;
+use Cake\Utility\Text;
+
+class Renderer
```
Comment (markstory) [inline src/Mailer/Renderer.php]: This should have a class doc. Is this class a public one that folks can extend/replace/rely on or is it intended to be an internal one?

File: src/Mailer/Renderer.php [hunk 3]
```diff
@@ -0,0 +1,462 @@
...
+     * @var \Cake\Mailer\Email
+     */
+    protected $email;
+
+    /**
+     * Constructor
+     */
+    public function __construct()
+    {
+        $this->appCharset = Configure::read('App.encoding');
```
Comment (markstory) [inline src/Mailer/Renderer.php]: Should this be a constructor argument?

File: src/Mailer/Renderer.php (hunk 2 above)
Comment (ADmad) [reply src/Mailer/Renderer.php]: We could make it internal. I personally can't think of a use case where someone might want to extend/modify it.

File: src/Mailer/Renderer.php (hunk 3 above)
Comment (ADmad) [reply src/Mailer/Renderer.php]: Probably. For now I had just copied from `Email`'s constructor.

File: src/Mailer/Renderer.php [hunk 4]
```diff
@@ -0,0 +1,462 @@
...
+            $attachments = $this->attachFiles($boundary);
+            $msg = array_merge($msg, $attachments);
+        }
+        if ($hasAttachments || $hasMultipleTypes) {
+            $msg[] = '';
+            $msg[] = '--' . $boundary . '--';
+            $msg[] = '';
+        }
+
+        return [$msg, $boundary, $textMessage, $htmlMessage];
```
Comment (ADmad) [inline src/Mailer/Renderer.php]: I should change this to associative array.

**Review by robertpustulka — CHANGES_REQUESTED** [review]
Could the view builder go to the renderer as well? It would make more sense to remove any rendering/templating stuff from the Email class.

If not possible, maybe we could split the Email class into Renderer and Message each with their own responsibilities and leave Email as a "facade" with BC API eventually dropping it in 5.x?

Ideally sending an email could look something like this:
```php
$renderer = new Renderer();
$renderer->setVar(...)->setTemplate(...);
$message = new Message();
$message->setFrom(...)->setTo(...)->setSubject()->setContent($renderer->getContent());

$transport->send($message);

//or hidden under BC Email "facade"
$email = new Email();
$email->setVar(...)->setTemplate(...)->setFrom(...)->setTo(...)->setSubject()->send();
```

**Comment by ADmad** [comment]
> Could the view builder go to the renderer as well? It would make more sense to remove any rendering/templating stuff from the Email class.

Done which proxying relevant Email methods to the Renderer's ViewBuilder. (I'll take care of the docblocks and stuff later).

> If not possible, maybe we could split the Email class into Renderer and Message each with their own responsibilities and leave Email as a "facade" with BC API eventually dropping it in 5.x?

That would be ideal but I don't think I can bring myself to put in the extra work to achieve that :slightly_smiling_face:.

**Review by dereuromark — CHANGES_REQUESTED** [review]
My comment has not been fixed yet. But yeah, can be done at the end.

**Comment by robertpustulka** [comment]
> Done, with proxying relevant Email methods to the Renderer's ViewBuilder. (I'll take care of the docblocks and stuff later).

OK, looks good for now.

**Review by robertpustulka — APPROVED** [review]

**Comment by ADmad** [comment]
> I can add Message class in another PR later.

Great, I'll wrap this up soon and pass on the baton to you :).

File: src/Mailer/Email.php [hunk 5]
```diff
@@ -1507,7 +1507,6 @@ public static function deliver(
         if (is_array($config) && !isset($config['transport'])) {
             $config['transport'] = 'default';
         }
-        /* @var \Cake\Mailer\Email $instance */
```
Comment (dereuromark) [inline src/Mailer/Email.php]: Are you sure you should remove this?
The discovery of setTo() method can not work without it.

File: src/Mailer/Renderer.php [hunk 6]
```diff
@@ -72,11 +72,13 @@ class Renderer
...
-     * Constructor
+     * Constructor.
+     *
+     * @param string|null $appCharset Application's character set.
      */
-    public function __construct()
+    public function __construct(?string $appCharset = null)
     {
-        $this->appCharset = Configure::read('App.encoding');
+        $this->appCharset = $appCharset ?? Configure::read('App.encoding');
```
Comment (dereuromark) [inline src/Mailer/Renderer.php]: In these cases it is probably better to use non cloaking `?:` check, but no big deal.

**Review by dereuromark — APPROVED** [review]

**Comment by ADmad** [comment]
I think everyone's comments are addressed now.

**Comment by markstory** [comment]
Looks good to me. This seems like a change that should be mentioned in the docs.

**Comment by ADmad** [comment]
Usage wise this doesn't make much difference to end users. Only thing is they have to use `setViewVars()` instead of `set()` for setting view vars. I'll open a docs PR to update code examples accordingly and mention the removed/added methods for `Email` on 4.0 migration page.

**Pull request merged**